drivers/spi/spi_silabs_eusart.c

```diff
@@ -152,6 +227,343 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	struct spi_silabs_eusart_data *instance_data = dev->data;
+	const struct spi_silabs_eusart_config *instance_conf = dev->config;
+
+	struct dma_config dma_cfg;
+
+	size_t current_total_transfer_size = 0U;
+	size_t padding_size = 0U;
+	size_t buffer_count = 0U;
+	size_t dma_block_config_idx = 0U;
+	int ret = 0U;
+
+	if (is_tx) {
+		current_buffer = &instance_data->ctx.current_tx[0];
+		buffer_count = instance_data->ctx.tx_count;
+		current_blk_cfg = &(instance_data->dma_channel_tx.dma_descriptors[0]);
+	} else {
+		current_buffer = &instance_data->ctx.current_rx[0];
+		buffer_count = instance_data->ctx.rx_count;
+		current_blk_cfg = &(instance_data->dma_channel_rx.dma_descriptors[0]);
+	}
+
+	/* For each buffer in the instance context buffer set. */
+	for (size_t i = 0U; i < buffer_count; i++) {
+		size_t buffer_remaining_count = current_buffer->len;
+
+		current_total_transfer_size += current_buffer->len;
+
+		if (buffer_remaining_count == 0U) {
+			/* Skip zero length buffers */
+			continue;
+		}
+
+		/* Create the required number of dma transfer block for this buffer. */
+		while (dma_block_config_idx < SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+			uint32_t index_in_buffer = current_buffer->len - buffer_remaining_count;
+
+			buffer_remaining_count = spi_silabs_build_dma_transfer_block(
+				instance_conf, &(current_blk_cfg[dma_block_config_idx]),
+				&(((uint8_t *)current_buffer->buf)[index_in_buffer]),
+				buffer_remaining_count, is_tx);
+
+			dma_block_config_idx++;
+
+			if (dma_block_config_idx > SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+				LOG_ERR("DMA channel transfer blocks exausted. Can't prepare the "
+					"transaction.");
+				return -ENOMEM;
+			}
+
+			if (buffer_remaining_count == 0U) {
+				current_buffer++;
+				break;
+			}
+		}
+	}
+
+	padding_size = total_padded_transaction_size - current_total_transfer_size;
+
+	while (padding_size) {
+
+		padding_size = spi_silabs_build_dma_transfer_block(
+			instance_conf, &current_blk_cfg[dma_block_config_idx], NULL, padding_size,
+			is_tx);
+
+		if (++dma_block_config_idx >= SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+			LOG_ERR("DMA transfer blocks exausted. Can't configure the transaction.");
+			return -ENOMEM;
+		}
+	}
+
+	spi_silabs_build_dma_config(dev, &dma_cfg,
+				    (const struct dma_block_config *)&current_blk_cfg[0],
+				    dma_block_config_idx, is_tx);
+
+	uint32_t dma_channel = 0U;
+	const struct device *dma_device = NULL;
+
+	if (is_tx) {
+		dma_channel = instance_data->dma_channel_tx.channel_nb;
+		dma_device = instance_data->dma_channel_tx.dma_device;
+		dma_cfg.dma_slot = instance_data->dma_channel_tx.dma_slot;
+	} else {
+		dma_channel = instance_data->dma_channel_rx.channel_nb;
+		dma_device = instance_data->dma_channel_rx.dma_device;
+		dma_cfg.dma_slot = instance_data->dma_channel_tx.dma_slot;
+	}
+
+	ret = dma_config(dma_device, dma_channel, &dma_cfg);
+
+	if (ret == -ENOMEM) {
+		LOG_ERR("DMA driver failed allocation.");
+		return ret;
+	}
+
+	return ret;
+}
```

Comment:
This function is definitely too long (100LoC) and too complex (5 level of indentation, 12 local variables). Normal humans can't juggle with more than 7 variables. Another symptom that this function is too complex: the length of the local variables.

Reply:
(I have just noticed the other reviewers did the same comment)